### PR TITLE
Go: add support for stream handling of attachments

### PR DESCRIPTION
**Pull request by wkalt**
Adds support for streaming reads and writes of attachment data. With this change, ParseAttachment is dropped as a public function and replaced with a private function, parseAttachmentReader.

A new type AttachmentReader is exposed, which exposes the Data portion of an attachment as a reader that can be streamed. The Attachment type is modified to include a new ContentLength field, to allow the writer to write the attachment length before stream processing the body.

A new option, AttachmentCallback, is exposed in the Lexer. Rather than emitting attachments, the Lexer now parses an attachment reader and calls the supplied callback (if supplied), and consumes any unconsumed bytes from the attachment record (whether or not a callback is supplied).

This is a breaking change in the Go SDK, and introduces some awkwardness due to the presence of the new ContentLength field in the Attachment record. In the specification, that value is folded into the Data field, however this is not a tenable approach for stream handling. The actual binary format is not affected by this - only the way in which the binary format is described.

This change is required for readers and writers to be able to handle attachments bigger than available RAM. Without it, attaching a sufficiently large attachment will crash the reader/writer with an OOM.

**Public-Facing Changes**

**Description**

### CODE REVIEW

**Comment by wkalt** [comment]
The point raised in the commit message about ContentLength relates to this issue: https://github.com/foxglove/mcap/issues/736

Before merging this patch we should pick what name we want to use for that field, if people are on board with the idea of splitting Data into ContentLength and Data. I don't feel too strongly about whether that spec change needs to go in before or after this merges, but before is probably slightly preferable.

Also whatever approach we go with for stream handling (a callback mechanism in this patch) should probably land in the other SDKs at some point, so we should make sure we are all comfortable with the approach.

File: go/mcap/lexer.go
Comment (james-rms) [inline go/mcap/lexer.go:225]: can we check if this is a Seeker and seek ahead instead if possible?

File: go/mcap/parse.go
Comment (james-rms) [inline go/mcap/parse.go:269]: Can we make CRC validation optional somehow?

File: go/mcap/parse.go
Comment (james-rms) [inline go/mcap/parse.go:290]: how do you feel about implementing a custom `LimitedReadSeeker` and returning that instead? The LimitReader source is not particularly complicated, it doesn't feel too hard to adapt.

File: go/mcap/mcap.go
Comment (james-rms) [inline go/mcap/mcap.go:246]: could alternatively cast this back to [LimitedReader](https://cs.opensource.google/go/go/+/go1.19.3:src/io/io.go;l=465) and check N

Comment (wkalt) [inline go/mcap/mcap.go:246]: this should just read one byte - that's all that's necessary to check if we need to fail

File: go/mcap/parse.go
Comment (wkalt) [reply go/mcap/parse.go:290]: what's the need for seeking?

Comment (wkalt) [reply go/mcap/parse.go:290]: also, I think that's going to require retention of the already-read data in memory if it doesn't exist on disk. Exposing io.Reader seems correct to me here; the limit aspect is just to ensure it EOFs at the right place.

File: go/mcap/parse.go
Comment (wkalt) [reply go/mcap/parse.go:269]: yeah, I can add the option to skip the computation of CRC. Actual validation of the CRC is up to the callback function the user supplies.

File: go/mcap/lexer.go
Comment (wkalt) [reply go/mcap/lexer.go:225]: yep, will do

File: go/mcap/mcap.go
Comment (wkalt) [reply go/mcap/mcap.go:246]: yeah - I think maybe the thing to do here is convert `Data` to a private `data` typed as a LimitReader, and then expose it through a `Data() io.Reader` method. This can then avoid the cast and otherwise do as you suggest. No need to read.

Comment (wkalt) [reply go/mcap/mcap.go:246]: same considerations should apply to the next function as well

File: go/mcap/lexer.go
Comment (wkalt) [inline go/mcap/lexer.go:478]: expand comment

File: go/mcap/lexer_test.go
Comment (wkalt) [inline go/mcap/lexer_test.go:367]: make some assertion here for clarity, should be nil or io.EOF

File: go/mcap/mcap.go
Comment (wkalt) [inline go/mcap/mcap.go:234]: group this with the other private fields

File: go/mcap/testutils.go
Comment (wkalt) [inline go/mcap/testutils.go:150]: seems like this should be 8. why doesn't this cause a parse failure? I guess because the test isn't doing anything with the CRC and is interpreting the CRC bytes as part of the data size. Since it's all zeros due to the empty content that would erroneously pass.

File: go/mcap/lexer.go
Comment (wkalt) [reply go/mcap/lexer.go]: added that

File: go/mcap/mcap.go
Comment (wkalt) [reply go/mcap/mcap.go]: added in both places

File: go/mcap/parse.go
Comment (james-rms) [reply go/mcap/parse.go]: If the underlying reader is a Seeker (which we already specialize for in some cases), then it may make sense to hand the user the ability to seek through the attachment for parts that they want.

Comment (wkalt) [reply go/mcap/parse.go]: I'm struggling to see a way to make this work that won't seem clunky or work against the CRC validation features.

* CRC validation seems like it would definitely break as soon as the user seeks around the attachment. If they don't read the whole thing, or read it in the wrong order, their CRC isn't going to validate.
* I don't think AttachmentReader should type the Data section with an io.ReadSeeker if there is any possibility that the thing cannot Seek (which is the case when doing a streaming read). So I think it would return an io.Reader, and if we got this working in some manner it would be up to the caller to attempt to cast to a ReadSeeker, and if successful seek away and break their own CRC validation.
* The brute force mechanism I think would be to make CRCReader a public type, and expose its underlying reader. Since the underlying reader is a LimitedReader that also exposes its underlying reader, by just making the CRCReader public the user would be able to call, effectively, `rs, ok := ar.Data().(CRCReader).R.(LimitedReader).R.(io.ReadSeeker)` and seek on that if successful. That would actually need to be split across multiple type assertions in case one of the intermediate ones fail. Definitely seems too clunky to me to ship.

Open to ideas on clean ways to handle this. I don't think it's a solution to just implement a LimitedReadSeeker and return that (even behind the io.Reader) because in common cases it would not actually be seekable. My inclination at the moment is to keep this as is and force the reader to consume the whole attachment (assuming they want any of it), and wait for feedback before optimizing. I think the usual case for dealing with attachments in mcap files is going to be to dump them out to disk for post-processing rather than do seeking reads on them in place - but we won't know till it gets some use.

If users have a Seekable reader they also have the option of addressing the attachment via the index and seeking around however they want, which the SDK fully supports. IMO that is probably enough to cover this usecase for seeking reads on an attachment in-place.

**Review by james-rms — APPROVED** [review]
This looks like it's going in a good direction - once it's possible to skip CRC calculation when reading attachments, I'm happy for this to merge.

**Pull request merged**
